File: backend/app/schemas/collection.py
"""集合树管理 API 的请求和响应结构。"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class CollectionCreate(BaseModel):
    name: str
    parent_id: str | None = "collection_root"
    description: str | None = Field(default=None, max_length=1000)


class CollectionUpdate(BaseModel):
    name: str | None = None
    description: str | None = Field(default=None, max_length=1000)


class CollectionMove(BaseModel):
    target_parent_id: str | None = "collection_root"


class CollectionMerge(BaseModel):
    target_collection_id: str


class CollectionDelete(BaseModel):
    reason: str | None = None


class CollectionRead(BaseModel):
    id: str
    parent_id: str | None
    name: str
    description: str | None = None
    path: str
    is_system: bool
    is_deleted: bool
    deletion_id: str | None = None
    direct_question_count: int = 0
    total_question_count: int = 0
    children: list["CollectionRead"] = []
    created_at: datetime | None = None
    updated_at: datetime | None = None


class CollectionDeletionRead(BaseModel):
    id: str
    root_collection_id: str
    collection_ids: list[str]
    question_ids: list[str]
    reason: str | None = None
    restored_at: datetime | None = None
    created_at: datetime | None = None

File: frontend/src/types/collection.ts
export const ROOT_COLLECTION_ID = "collection_root";
export const UNFILED_COLLECTION_ID = "collection_unfiled";

export type CollectionNode = {
  id: string;
  parent_id: string | null;
  name: string;
  description?: string | null;
  path: string;
  is_system: boolean;
  direct_question_count: number;
  total_question_count: number;
  /** Present only for a deleted collection; required by the restore endpoint. */
  deletion_id?: string | null;
  is_deleted: boolean;
  children: CollectionNode[];
};

export type CollectionCreatePayload = { name: string; parent_id?: string | null; description?: string | null };
export type CollectionUpdatePayload = Partial<Pick<CollectionCreatePayload, "name" | "description">>;
export type CollectionDeletePayload = { reason?: string | null };

export type CollectionDeletion = {
  id: string;
  root_collection_id: string;
  collection_ids: string[];
  question_ids: string[];
  reason?: string | null;
  restored_at?: string | null;
  created_at?: string | null;
};

export type QuestionPlacement = { question_id: string; collection_id: string | null };
export type BulkMoveResponse = {
  moved_count?: number;
  failed_count?: number;
  moved_ids?: string[];
  failures?: Array<{ question_id: string; message: string }>;
};
